### ISSUE TITLE
Consider building fortune-backend in docker

### ISSUE THREAD

**Issue by alekssaul** [azwickey-pivotal/pks-walkthrough#2]
Docker (since 1.12) support multi-staged container builds, such that you can build the jar in docker container with java and move the final artifact to a runtime container.

This would allow building the artifact/final container on machines without maven/java installed.

You can see the documentation in; https://docs.docker.com/engine/userguide/eng-image/multistage-build/ I have an example in `https://github.com/alekssaul/golang_helloworld/blob/master/Dockerfile`.

I can file a PR when I get some time.

**Comment by azwickey-pivotal** [comment]
This will considerably increase the size of the image as you are downloading a lot of assets into the image.  Additionally, its a best practice to separate your artifacts from your containers as you want to version them separately.  E.G. You should be able to patch the container w/o changing your code artifact.

**Comment by alekssaul** [comment]
Makes sense, coming from a docker centric world wanted to give a sense of my thought process.. 

Essentially the "build" container solves the problem of seperating the artifact from your runtime container such that:
- build container is versioned and immutable (some caveats to this statement when relying on docker tags, but docker image hashes should be immutable)
- build asset/source code is immutable via git sha

resulting in a system where artifact can be reproduced the same without relying on CI system (which can be hard to keep "clean" in case of jenkins etc). In CI systems (outside concourse) there is the added benefit of reproducibility since artifact generation does not rely on CI system builder's toolchain. This may come into play in case of a artifact management system failure. 

The end artifact (runnable docker container image) size should not change, however during the build process you most likely will have larger image pulls as you will have to download a container image for `build` vs `run` stages. 

Since the stages (build vs run) are distinct, you can patch the runtime environment without changing the build environment.
